### PR TITLE
raft: add learner

### PR DESCRIPTION
**Pull request by shuaili87**
Partly implement https://github.com/coreos/etcd/issues/8568,  only add Nonvoter support.

for protobuf, maybe we can add a version for pb.ConfState.

### CODE REVIEW

**Comment by javaforfun** [comment]
@xiang90 @siddontang

**Comment by xiang90** [comment]
sorry for the delay. i will try to give this a review this week.

File: raft/raftpb/raft.proto [hunk 1]
```diff
@@ -75,14 +75,26 @@ message HardState {
...
+
+message Server {
+	optional uint64        node     = 1 [(gogoproto.nullable) = false];
+	optional SuffrageState suffrage = 2 [(gogoproto.nullable) = false];
+}
+
 message ConfState {
-	repeated uint64 nodes = 1;
+	repeated uint64 nodes   = 1;
+	repeated Server servers = 2;
```
Comment (xiang90) [inline raft/raftpb/raft.proto]: can we make this simpler by just having

```
repeated uint64 nodes = 1;
repeated uint64 learners= 2;
```
?

**Comment by xiang90** [comment]
can we rename Nonvoter to Learners?

also i would like to simplify the code path and minimize code change by having an explicit list of learners instead of a mixed list of Servers (which can be peer or non-voter).

File: raft/raftpb/raft.proto (hunk 1 above)
Comment (javaforfun) [reply raft/raftpb/raft.proto]: we may also have staging nodes after addVoter feature

Comment (xiang90) [reply raft/raftpb/raft.proto]: i do not like the idea of staging server. i would rather let the application itself promote the leader to a node explicitly.

Comment (xiang90) [reply raft/raftpb/raft.proto]: basically, the less raft layer needs to care about the better.

**Comment by xiang90** [comment]
the learner should always stay as follower state, and never try to promote or vote, right? where do we prevent that from happening?

**Comment by javaforfun** [comment]
> the learner should always stay as follower state, and never try to promote or vote, right? where do we prevent that from happening?

the promotable will always false for learner:
```
func (r *raft) promotable() bool {
 	_, ok := r.prs[r.id]
	return ok && !r.isLearner
 }
```
learner should be able to vote. because if we addNode a learner,  the voters may think that node is voter,  but that node may think its still learner.
for example,  
1. cluster has [A, B, C], with A is leader,
2. addLearner D.  the majority is still 2.
3. addNode D. [A, B, C] both apply the confchange, but D don't recv the append log due to log lag or network jitter.
4. A down. [B, C] is voter, they think D is voter too, the majority is 3.  but D think itself is learner.
if learner can't vote,  the cluster will hang until A recover.  if learner can vote, the cluster will ok immediately.

**Comment by xiang90** [comment]
> learner should be able to vote. because if we addNode a learner, the voters may think that node is voter, but that node may think its still learner.

The same thing happens with normal add node case. There are two motivations for adding learner:

1. scale up read request when consistency can be relaxed
2. avoid slow member to disrupt the cluster (only promote to normal member when it can catch up)

What you are trying to do is to enhance 2. I would like to know why 2 itself is not strong enough. 

Do you have a specific use case or failure mode in mind?

**Comment by javaforfun** [comment]
In our case, we have 3 nodes in one raft group.  If we want replace one node,  we will add a new node first, then remove the old one.
But if a node down,  the replace operation can't run as usual.  Because if we add a new node first, the majority will be 3, and client requests will hang/timeout until the new node catch up.

**Comment by siddontang** [comment]
Assume we have 3 nodes, A, B and C, and want to replace A with D, so now we can:

1. Add learner D
2. Remove A
3. Add D

The corner case here is after 2, we have only B, C and if any one of them fails, the cluster can't work. 

If the learner D can vote, this is not a problem, but here we have another corner case, after 1, we have A, B, C, D, A and D are on the same node, so if the node crashes, the cluster can't work too.

We have the same problem when we do:

1. Add learner D
2. Add D
3. Remove A

I don't think the learner mechanism can fix this corner case, maybe only Raft Membership Change can do. So I suggest keeping things simple and letting learner don't be able to vote.

**Comment by xiang90** [comment]
I would say keep it as simple as possible for now. if the corner case does happen in reality or if we have a more concrete real world use case and failure case, we can improve it.

**Comment by xiang90** [comment]
Just to be more clear, after adding the learner concept, we already largely improve the reliability of membership change process. We will only promote a learner to a member if it can catch up.

File: raft/raft.go [hunk 2]
```diff
@@ -504,9 +531,11 @@ func (r *raft) bcastHeartbeatWithCtx(ctx []byte) {
 // r.bcastAppend).
 func (r *raft) maybeCommit() bool {
 	// TODO(bmizerany): optimize.. Currently naive
-	mis := make(uint64Slice, 0, len(r.prs))
-	for id := range r.prs {
-		mis = append(mis, r.prs[id].Match)
+	mis := make(uint64Slice, 0, r.voterCount())
```
Comment (siddontang) [inline raft/raft.go]: seem we can still use `len(r.prs)` to reduce the extra learner check in `voterCount`.

**Comment by siddontang** [comment]
PTAL @xiang90

**Comment by xiang90** [comment]
@siddontang i would like to take another look after @javaforfun disables the vote of the learner.

**Comment by siddontang** [comment]
Hi @xiang90 

Seem the learner can't vote now, see https://github.com/coreos/etcd/pull/8605/files#diff-a8629a0eb6da1c6c3805111f1f1af85eR376

**Comment by javaforfun** [comment]
@xiang90 If learner can't vote, how to prevent this problem?  https://github.com/coreos/etcd/pull/8605#issuecomment-333350492

**Comment by xiang90** [comment]
@javaforfun we are not going to solve the problem you brought up for now. we should focus on solving:

1. scale up read request when consistency can be relaxed
2. avoid slow member to disrupt the cluster (only promote to normal member when it can catch up)

**Comment by siddontang** [comment]
I agree with @xiang90, here we don't need a 100% solution for ConfChange.

**Comment by siddontang** [comment]
PTAL @xiang90

File: raft/raft.go [hunk 3]
```diff
@@ -116,6 +116,8 @@ type Config struct {
 	// used for testing right now.
 	peers []uint64
 
+	learners []uint64
```
Comment (xiang90) [inline raft/raft.go]: doc string?

File: raft/raft.go [hunk 4]
```diff
@@ -288,19 +292,22 @@ func newRaft(c *Config) *raft {
 	if err != nil {
 		panic(err) // TODO(bdarnell)
 	}
-	peers := c.peers
-	if len(cs.Nodes) > 0 {
-		if len(peers) > 0 {
+	voters := c.peers
```
Comment (xiang90) [inline raft/raft.go]: use peers instead of voters. i do not want to create a new term for now.

File: raft/raft.go [hunk 5]
```diff
@@ -313,8 +320,17 @@ func newRaft(c *Config) *raft {
 		readOnly:                  newReadOnly(c.ReadOnlyOption),
 		disableProposalForwarding: c.DisableProposalForwarding,
 	}
-	for _, p := range peers {
-		r.prs[p] = &Progress{Next: 1, ins: newInflights(r.maxInflight)}
+	for _, n := range voters {
+		r.prs[n] = &Progress{Next: 1, ins: newInflights(r.maxInflight), isLearner: false}
```
Comment (xiang90) [inline raft/raft.go]: revert this change. bool is false by default.

File: raft/raftpb/raft.proto [hunk 6]
```diff
@@ -76,13 +76,15 @@ message HardState {
...
-	repeated uint64 nodes = 1;
+	repeated uint64 nodes    = 1;  // Voters
+	repeated uint64 learners = 2;  // Nonvoters
 }
 
 enum ConfChangeType {
 	ConfChangeAddNode    = 0;
 	ConfChangeRemoveNode = 1;
 	ConfChangeUpdateNode = 2;
+	ConfChangeAddLearner = 3;
```
Comment (xiang90) [inline raft/raftpb/raft.proto]: ConfChangeAddLearnerNode

File: raft/raft.go [hunk 7]
```diff
@@ -313,8 +320,17 @@ func newRaft(c *Config) *raft {
...
+	}
+	for _, n := range learners {
+		if _, has := r.prs[n]; has {
+			panic(fmt.Sprintf("cannot specify both Voter and Learner for node: %x", n))
+		}
+		r.prs[n] = &Progress{Next: 1, ins: newInflights(r.maxInflight), isLearner: true}
```
Comment (xiang90) [inline raft/raft.go]: maybe it is easier to have a separate prs map for learners.

File: raft/raft.go [hunk 8]
```diff
@@ -1171,33 +1213,58 @@ func (r *raft) restore(s pb.Snapshot) bool {
 
 	r.raftLog.restore(s)
 	r.prs = make(map[uint64]*Progress)
-	for _, n := range s.Metadata.ConfState.Nodes {
+	r.restoreNode(s.Metadata.ConfState.Nodes, false)
+	r.restoreNode(s.Metadata.ConfState.Learners, true)
+	return true
+}
+
+func (r *raft) restoreNode(nodes []uint64, isLearner bool) {
```
Comment (xiang90) [inline raft/raft.go]: probably restoreNode can take both nodes and learnerNodes as args.

File: raft/raft.go [hunk 9]
```diff
@@ -1171,33 +1213,58 @@ func (r *raft) restore(s pb.Snapshot) bool {
...
-		r.setProgress(n, match, next)
+		r.setProgress(n, match, next, isLearner)
 		r.logger.Infof("%x restored progress of %x [%s]", r.id, n, r.prs[n])
 	}
-	return true
 }
 
 // promotable indicates whether state machine can be promoted to leader,
-// which is true when its own id is in progress list.
+// which is true when its own id is in progress list and its not learner.
```
Comment (xiang90) [inline raft/raft.go]: it is not a learner

File: raft/raft.go [hunk 10]
```diff
@@ -1171,33 +1213,58 @@ func (r *raft) restore(s pb.Snapshot) bool {
...
 func (r *raft) promotable() bool {
 	_, ok := r.prs[r.id]
```
Comment (xiang90) [inline raft/raft.go]:
```go
if r.isLearer {
    return false
}
...
```
is cleaner

File: raft/raft.go [hunk 11]
```diff
@@ -1171,33 +1213,58 @@ func (r *raft) restore(s pb.Snapshot) bool {
...
 
 func (r *raft) addNode(id uint64) {
+	r.addLearnerNode(id, false)
+}
+
+func (r *raft) addLearner(id uint64) {
+	r.addLearnerNode(id, true)
+}
+
+func (r *raft) addLearnerNode(id uint64, isLearner bool) {
```
Comment (xiang90) [inline raft/raft.go]: addNodeOrLearnerNode. 

or probably just duplicate code a little bit, and remove this func.

File: raft/raft.go [hunk 12]
```diff
@@ -1171,33 +1213,58 @@ func (r *raft) restore(s pb.Snapshot) bool {
...
-		// Ignore any redundant addNode calls (which can happen because the
-		// initial bootstrapping entries are applied twice).
-		return
+		if r.prs[id].isLearner == isLearner {
+			// Ignore any redundant addNode calls (which can happen because the
+			// initial bootstrapping entries are applied twice).
+			return
+		}
+		if isLearner {
+			// can only change Learner to Voter
```
Comment (xiang90) [inline raft/raft.go]: log here.

File: raft/raft.go [hunk 13]
```diff
@@ -1171,33 +1214,62 @@ func (r *raft) restore(s pb.Snapshot) bool {
...
+			r.logger.Infof("%x ignore addLearner for %x [%s]", r.id, id, r.prs[id])
```
Comment (xiang90) [inline raft/raft.go]: need to log the reason why addLearner is ignored.

File: raft/raft.go [hunk 14]
```diff
@@ -346,7 +363,18 @@ func (r *raft) hardState() pb.HardState {
...
+	count := 0
+	for _, p := range r.prs {
+		if !p.isLearner {
+			count++
+		}
+	}
+	return count
+}
+
+func (r *raft) quorum() int { return r.voterCount()/2 + 1 }
```
Comment (siddontang) [inline raft/raft.go]: This function is called frequently, maybe now we can use a variable to cache the quorum.

Comment (siddontang) [reply raft/raft.go]: @xiang90

Comment (xiang90) [reply raft/raft.go]: agreed

File: raft/raft_test.go [hunk 15]
```diff
@@ -3278,7 +3388,7 @@ func entsWithConfig(configFunc func(*Config), terms ...uint64) *raft {
 	for i, term := range terms {
 		storage.Append([]pb.Entry{{Index: uint64(i + 1), Term: term}})
 	}
-	cfg := newTestConfig(1, []uint64{}, 5, 1, storage)
+	cfg := newTestConfig(1, []uint64{1, 2, 3, 4, 5}, 5, 1, storage)
```
Comment (xiang90) [inline raft/raft_test.go]: why this is changed?

File: raft/raft_test.go [hunk 16]
```diff
@@ -3293,7 +3403,7 @@ func entsWithConfig(configFunc func(*Config), terms ...uint64) *raft {
 func votedWithConfig(configFunc func(*Config), vote, term uint64) *raft {
 	storage := NewMemoryStorage()
 	storage.SetHardState(pb.HardState{Vote: vote, Term: term})
-	cfg := newTestConfig(1, []uint64{}, 5, 1, storage)
+	cfg := newTestConfig(1, []uint64{1, 2, 3, 4, 5}, 5, 1, storage)
```
Comment (xiang90) [inline raft/raft_test.go]: why this is changed?

File: raft/raftpb/raft.proto [hunk 17]
```diff
@@ -76,13 +76,15 @@ message HardState {
 }
 
 message ConfState {
-	repeated uint64 nodes = 1;
+	repeated uint64 nodes    = 1;  // Voters
```
Comment (xiang90) [inline raft/raftpb/raft.proto]: remove this comment. `voter` is meaningless without any context. or we need to explain exactly what is nodes.

File: raft/raftpb/raft.proto [hunk 18]
```diff
@@ -76,13 +76,15 @@ message HardState {
...
+	repeated uint64 learners = 2;  // Nonvoters
```
Comment (xiang90) [inline raft/raftpb/raft.proto]: same here. we need to explain what it nonvoters.

**Comment by shuaili87** [comment]
```
maybe it is easier to have a separate prs map for learners.
```
we may need to handle the prs before merging this PR. also the prs in status struct, https://github.com/coreos/etcd/blob/master/raft/status.go#L30
```go
type Status struct {
	ID uint64

	pb.HardState
	SoftState

	Applied  uint64
	Progress map[uint64]Progress

	LeadTransferee uint64
}
```

File: raft/raft.go (hunk 8 above)
Comment (siddontang) [reply raft/raft.go]: +1

**Comment by siddontang** [comment]
> maybe it is easier to have a separate prs map for learners.

Seem this can handle leaner more easily, but may introduce complexity in other places like in `bcast`.

**Comment by xiang90** [comment]
@siddontang 

right. someone should give it a try at least to see if it is simpler.

**Comment by xiang90** [comment]
closed in favor of #8751
